src/items.rs

```diff
@@ -656,8 +656,7 @@ impl<'a> FmtVisitor<'a> {
         }
 
         let context = self.get_context();
-        // 1 = ','
-        let shape = self.shape().sub_width(1)?;
+        let shape = self.shape();
```

Comment:
The `sub_width(1)` is still important for rewriting each enum variant as we need to take the trailing comma into account. The bug here is that we're apply `sub_width(1)` prematurely. We still need to reduce the width by 1 **after** we rewrite the attributes as the trailing comma doesn't impact the attributes.

Additionally, I think it would be best if these changes were [version gated](https://github.com/rust-lang/rustfmt/blob/master/Contributing.md#version-gate-formatting-changes).

probably best to change this to something like:

```rust
let attrs_str = if context.config.version() == Version::Two {
    field.attrs.rewrite(&context, shape)?
} else {
   // Version::One formatting that was off by 1. See issue #5801
   field.attrs.rewrite(&context, shape.sub_width(1)?)?
};

// sub_width(1) to take the trailing comma into account
let shape = shape.sub_width(1)?;
```

Reply:
wait, this is only an issue when `wrap_comments=true` is set, right? In that case we don't need the version gate since `wrap_comments=true` is an unstable option. I'd then expect the code change to look something like:

```rust
let attrs_str = field.attrs.rewrite(&context, shape)?;
// sub_width(1) to take the trailing comma into account
let shape = shape.sub_width(1)?;
```

Reply:
As I was thinking about this more I'm now 100% convinced this change needs to be version gated as attribute formatting is also influenced by stable configuration options like `attr_fn_like_width` and `use_small_heuristics`.

Consider this extended example from what you have in the test case.

```rust
pub enum Severity {
    /// But here, this comment is 120 columns wide and the formatter wants to split it up onto two separate lines still.
    Error,
    /// This comment is 119 columns wide and works perfectly. Lorem ipsum. lorem ipsum. lorem ipsum. lorem ipsum lorem.
    Warning,
    #[something(AAAAAAAAAAAAA, BBBBBBBBBBBBBB, CCCCCCCCCCCCCCCC, DDDDDDDDDDDDD, EEEEEEEEEEEE, FFFFFFFFFFF, GGGGGGGGGGG)]
    AttrsWillWrap,
    #[something_else(hhhhhhhhhhhhhhhh, iiiiiiiiiiiiiiii, jjjjjjjjjjjjjjj, kkkkkkkkkkkkk, llllllllllll, mmmmmmmmmmmmmm)]
    AttsWontWrap,
}
```

Using a recent nightly `rustfmt 1.7.0-nightly (89e2160c 2023-12-27)` with the input above and these configuration options:

```toml
comment_width = 120
max_width = 120
wrap_comments = true
version = "One"
attr_fn_like_width = 120
```

produces this formatting:

```rust
pub enum Severity {
    /// But here, this comment is 120 columns wide and the formatter wants to split it up onto two separate lines
    /// still.
    Error,
    /// This comment is 119 columns wide and works perfectly. Lorem ipsum. lorem ipsum. lorem ipsum. lorem ipsum lorem.
    Warning,
    #[something(
        AAAAAAAAAAAAA,
        BBBBBBBBBBBBBB,
        CCCCCCCCCCCCCCCC,
        DDDDDDDDDDDDD,
        EEEEEEEEEEEE,
        FFFFFFFFFFF,
        GGGGGGGGGGG
    )]
    AttrsWillWrap,
    #[something_else(hhhhhhhhhhhhhhhh, iiiiiiiiiiiiiiii, jjjjjjjjjjjjjjj, kkkkkkkkkkkkk, llllllllllll, mmmmmmmmmmmmmm)]
    AttsWontWrap,
}
```

Reply:
I've updated the code to include the version gate. Would you also like me to update the test cases to match what you've as an example here?